tensorflow/c/kernels/summary_op_test.cc

```diff
@@ -0,0 +1,172 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/core/framework/attr_value.pb.h"
+#include "tensorflow/core/framework/attr_value_util.h"
+#include "tensorflow/core/framework/node_def.pb.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/platform/test.h"
+#include "tensorflow/core/framework/summary.pb.h"
+
+namespace tensorflow {
+namespace {
+
+class DummyDevice : public DeviceBase {
+ public:
+  explicit DummyDevice(Env* env) : DeviceBase(env) {}
+  Allocator* GetAllocator(AllocatorAttributes /*attr*/) override {
+    return cpu_allocator();
+  }
+};
+
+// Helper for comparing ouput and expected output
+static void EXPECT_SummaryMatches(const Summary& actual,
+                                  const string& expected_str) {
+  Summary expected;
+  (protobuf::TextFormat::ParseFromString(expected_str, &expected));
+  EXPECT_EQ(expected.DebugString(), actual.DebugString());
+}
+
+
+void TestScalarSummaryOp(Tensor* tags, Tensor* values, string expected_summary,
+                         error::Code expected_code) {
+  // Initialize node used to fetch OpKernel 
+  Status status;
+  NodeDef def;
+  def.set_op("ScalarSummary");
+
+  def.set_device(DEVICE_CPU);
+
+  AttrValue valuesTypeAttr;
+  SetAttrValue(values->dtype(), &valuesTypeAttr);
+  (*def.mutable_attr())["T"] = valuesTypeAttr;
+
+  def.add_input(
+      strings::StrCat("input1: ", DataTypeString(tags->dtype())));
+  def.add_input(
+    strings::StrCat("input2: ", DataTypeString(values->dtype())));
+
+  std::unique_ptr<OpKernel> kernel =
+      CreateOpKernel(DeviceType(DEVICE_CPU), nullptr, 
+      nullptr, def, 1, &status);
+  ASSERT_TRUE(status.ok()) << status.ToString();
+  OpKernelContext::Params params;
+  DummyDevice dummy_device(nullptr);
+  params.device = &dummy_device;
+  params.op_kernel = kernel.get();
+  AllocatorAttributes alloc_attrs; 
+  params.output_attr_array = &alloc_attrs;
+  gtl::InlinedVector<TensorValue, 4> inputs;
+  inputs.emplace_back(tags);
+  inputs.emplace_back(values);
+  params.inputs = &inputs;
+  OpKernelContext ctx(&params, 1); 
+  kernel->Compute(&ctx);
+  ASSERT_EQ(expected_code, ctx.status().code());
+  if (expected_code == error::OK){ 
+    Summary summary; 
+    ParseProtoUnlimited(&summary, ctx.mutable_output(0)->scalar<tstring>()());
```

Comment:
Nit: You can use ParseProtoUnlimited's void* + size_t overload instead to avoid a tstring->string copy that's occurring here.

Reply:
Oops, ignore this comment, it looks like there's an overload that accepts a tstring that already does this :)
https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/platform/protobuf.h;l=59;drc=0078fff2cc604a755d42650eeb3c0e4099df66a2